client/src/model/component/CircuitDiagram.hx

```diff
@@ -215,4 +215,124 @@ class CircuitDiagram extends Observable implements CircuitDiagramI implements Ob
         }
         return worldPointArray;
     }
+	
+	public function normalise() : Void {
+		// For all pairs of connections {c0, c1} such that c0 != c1 and c0 is close to c1
+		// if and both contain a port,
+			// create a link to connect them
+		// else one or both don't contain a port
+			// if c0 contains a port, swap c0 and c1
+			// move everything in c0 to c1
+		var connectionArray = new Array<Connection>();
+		for (link in this.linkArray) {
+			connectionArray.push(link.get_endpoint(0).getConnection());
+			connectionArray.push(link.get_endpoint(1).getConnection());
+		}
+		
+		for (component in this.componentArray) {
+			var ports = component.get_ports();
+			for (port in ports) {
+				connectionArray.push(port.getConnection());
+			}
+		}
+		
+		var markedConnections = new Array<Connection>();
+		for (connection in connectionArray) {
+			var otherConnections = connectionArray;
+			otherConnections.remove(connection);
+			for (otherConnection in otherConnections) {
+				var connectionDistance = Link.pointsDistance(connection.get_xPosition(), connection.get_yPosition(), otherConnection.get_xPosition(), otherConnection.get_yPosition());
+				if (connectionDistance <= 15) {
+					// here 15 is what is being considered as "close" to each other
+					if (connection.aPortIsConnecte() && otherConnection.aPortIsConnecte()) {
+						this.addLink(new Link(this, connection.get_xPosition(), connection.get_yPosition(), otherConnection.get_xPosition(), otherConnection.get_yPosition()));
+					} else {
+						if (connection.aPortIsConnecte()) {
+							for (element in connection.get_connectedElements()) {
+								otherConnection.connect(Std.downcast(element, Connectable));
+							}
+							markedConnections.push(connection);
+						}
+					}
+					
+					
+				} 
+			}
+		}
+		
+		// removing marked connections. Unclear is swapping connections in use cases is actually what it is supposed to be.
+		// removing such unused connections makes more sense. CHECK ?
+		for (connection in markedConnections) {
+			connectionArray.remove(connection);
+		}
+		
+		// For all connections c and for all links x not connected to c
+		// 		If c is not close to either endpoint of x, but is close to x
+        // 			Split x into two links each with an endpoint added to c
+		// VERIFY :: Can cause Short circuit!!
+		// this carries on with the remaining links
+		var removeLinks = new Array<Link>();
+		for (connection in connectionArray) {
+			for (link in this.linkArray) {
+				if ((link.get_endpoint(0).getConnection() != connection) && (link.get_endpoint(1).getConnection() != connection)) {
+					if (Link.pointToLine(connection.get_xPosition(), connection.get_yPosition(), link.get_endpoint(0).get_xPosition(), link.get_endpoint(0).get_yPosition(), link.get_endpoint(1).get_xPosition(), link.get_endpoint(1).get_yPosition()) <= 15) {
```

Comment:
Maybe we could make a method in Link to say how close a point is to it.

Reply:
I think there is a method called `pointToLine()`

Reply:
I think this method already does that, it might look ugly though.